text/trace/0220-messaging-semantic-conventions-span-structure.md

```diff
@@ -0,0 +1,482 @@
+
+For each producer scenario, a "Publish" span needs to be created. This span
+measures the duration of the call or operation that provides messages for
+sending or publishing to an intermediary. This call or operation (and the
+related "Publish" span) can either refer to a single message or a batch of
+multiple messages.
+
+It is recommended to create a "Create" span for every single message. "Create"
+spans can be created during the "Publish" operation as children of the
+"Publish" span. Alternatively, "Create" spans can be created independently of
+the "Publish" operation. In that case, SDKs may provide mechanisms to allow
+attaching independent contexts with messages.
+
+If a "Create" span exists for a message, its context must be attached to the
+message. If no "Create" span exists for a message, the context of the related
+"Publish" span must be attached to the message.
+
+### Consumer
+
+For many use cases, it is not possible to rely on the presence of "Process"
+spans for correlating producer with consumer traces: there are cases where a
+dedicated processing operation cannot be identified, or where processing
+happens in a different trace. Furthermore, processing operations often are not
+covered by messaging libraries and SDKs, but take place in application code.
+Consistently creating spans for "Processing" operations would require either
+effort from the application owner to correctly instrument those operations, or
+additional capabilities of messaging libraries and SDKs (e. g. hooks for
+processing callbacks, which can then be instrumented by the libraries or SDKs).
+
+While it is possible to create "Process" spans and correlate those with
+consumer traces in certain cases, this is not something that can be generally
+required. Therefore, it is more feasible to require the creation of "Deliver"
+spans (for push-based APIs) or "Receive" spans (for pull-based APIs) to
+correlate producer with consumer traces.
+
+#### Instrumenting push-based scenarios
+
+In push-based consumer scenarios, the delivery of messages is not initiated by
+the application code. Instead, callbacks or handlers are registered and then
+called by messaging SDKs to forward messages to the application.
+
+A "Deliver" span covers the call of such a callback or handler and should link
+to the "Create" spans of all messages that are forwarded via the respective
+call. Depending on the use case, "Deliver" spans can correlate with "Process"
+spans or other spans modelling processing operations.
+
+#### Instrumenting pull-based scenarios
+
+In pull-based consumer scenarios, the delivery of messages is requested by the
+application code. This usually involves a blocking call, which returns zero or
+more messages on completion.
+
+A "Receive" span covers such calls and should link to the "Create" spans of all
+messages that are forwarded via the respective call. Depending on the use case,
+"Receive" spans can correlate with "Process" spans or other spans modelling
```

Comment:
I think correlate is too ambiguous and we should avoid it whenever possible. Should it be either parent, link or both?

Reply:
We are ambiguous on purpose here. We don't specify any conventions for "Process" spans in this document (yet). Users might create those spans, but we leave it open whether they are created at all and how they are "correlated" to other spans.

Reply:
It will save time to a lot of people if we take a decision on this regard.

Reply:
We refrained from adding conventions for "Process" spans for several reasons:
* It's hard to define what "processing" a message entails. It might be a single operation, it also might be multiple operations, sometimes messages might be received but not be processed at all. As there are many ways of how to process a message, we don't want to limit users by any strong requirements.
* One aim of the conventions is, that they can be implemented by messaging SDKs, without the need for users to add additional instrumentation. Once we require the creation of process spans, this requirement will be very hard (or in many cases even impossible) to satisfy by SDK instrumentation.
* Given the above considerations, it is definitely possible to add recommendations about creating process spans now or at a future point in time, maybe in v2 of these conventions. As there are many problems to tackle, we thought it's preferable to get a smaller set of conventions stable earlier instead of taking longer for a comprehensive set of conventions.

Reply:
While I understand that a `Process` span is difficult to produce at an instrumentation level, I think it should still be defined if only to be marked as optional.  If not, then when they are added it will be done inconsistently.

Reply:
This was discussed in today's messaging workgroup. There's interest and value in adding recommendations about "Process" spans, however, for this OTEP we'd like to focus on the _minimum_ of spans and correlations that is required and backends can rely on and that can be provided by auto-instrumentation (at least one of "create" or "publish" on producer side, and at least one of "deliver" or "receive" on the consumer side). We'd like to keep this OTEP at this scope, as it already has a length that makes it challenging to review.

@tylerbenson Do you think that what's proposed in this OTEP can serve as a basis for adding recommendations about "process" spans? Would you be willing to come up with a proposal for adding "process" spans? Either based on this OTEP, or if that's not possible, highlight the conflicts you see?

Reply:
I agree that the process span definition should be done independently as this effort is already sufficiently large and complex. I am putting more focus on the FAAS/Java SIGs at the moment so I don't have cycles to contribute a proposal for the process span, but happy to review when one is ready.